File: examples/17-sagas/python/saga-orchestrator/main.py
import json
import logging
import os
import uuid
from contextlib import asynccontextmanager
from typing import Optional

import asyncpg
from fastapi import FastAPI
from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from pydantic import BaseModel, Field

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("saga")

DATABASE_URL = os.getenv("DATABASE_URL", "postgres://appuser:apppass@postgres:5432/appdb")
OTEL_ENDPOINT = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "http://lgtm:4318")
SHIPPING_FAIL = os.getenv("SHIPPING_FAIL", "false").lower() == "true"

resource = Resource.create({"service.name": "saga-orchestrator"})
provider = TracerProvider(resource=resource)
provider.add_span_processor(BatchSpanProcessor(OTLPSpanExporter(endpoint=f"{OTEL_ENDPOINT}/v1/traces")))
trace.set_tracer_provider(provider)
tracer = trace.get_tracer("saga")

db_pool: Optional[asyncpg.Pool] = None

STEPS = [
    {"name": "charge_payment", "compensate": "refund_payment"},
    {"name": "reserve_stock", "compensate": "release_stock"},
    {"name": "book_shipping", "compensate": "cancel_shipping"},
]


async def execute_step(step_name: str, context: dict) -> dict:
    if step_name == "charge_payment":
        payment_id = f"pay-{uuid.uuid4().hex[:8]}"
        log.info("charged payment %s for order %s", payment_id, context.get("order_id"))
        return {"payment_id": payment_id, "amount": context.get("total", 0)}

    if step_name == "reserve_stock":
        reservation_id = f"rsv-{uuid.uuid4().hex[:8]}"
        log.info("reserved stock %s for sku %s", reservation_id, context.get("sku"))
        return {"reservation_id": reservation_id, "sku": context.get("sku")}

    if step_name == "book_shipping":
        if SHIPPING_FAIL or context.get("fail_shipping"):
            raise RuntimeError("shipping service unavailable")
        shipment_id = f"shp-{uuid.uuid4().hex[:8]}"
        log.info("booked shipping %s", shipment_id)
        return {"shipment_id": shipment_id}

    if step_name == "refund_payment":
        log.info("refunded payment %s", context.get("charge_payment", {}).get("payment_id"))
        return {"refunded": True}

    if step_name == "release_stock":
        log.info("released stock %s", context.get("reserve_stock", {}).get("reservation_id"))
        return {"released": True}

    if step_name == "cancel_shipping":
        log.info("cancelled shipping %s", context.get("book_shipping", {}).get("shipment_id"))
        return {"cancelled": True}

    return {}


async def advance(saga_id: str):
    async with db_pool.acquire() as conn:
        async with conn.transaction():
            row = await conn.fetchrow(
                "SELECT id, status, step_index, context FROM sagas WHERE id=$1 FOR UPDATE",
                saga_id,
            )
            if not row or row["status"] != "RUNNING":
                return

            step_index = row["step_index"]
            context = json.loads(row["context"])

            if step_index >= len(STEPS):
                await conn.execute(
                    "UPDATE sagas SET status='COMPLETED', updated_at=now() WHERE id=$1",
                    saga_id,
                )
                return

            step = STEPS[step_index]
            with tracer.start_as_current_span(f"saga.{step['name']}") as span:
                span.set_attribute("saga.id", saga_id)
                span.set_attribute("saga.step", step["name"])
                try:
                    result = await execute_step(step["name"], context)
                    context[step["name"]] = result

                    await conn.execute(
                        "INSERT INTO saga_log (saga_id, step, action, result) VALUES ($1, $2, 'execute', $3)",
                        saga_id, step["name"], json.dumps(result),
                    )
                    await conn.execute(
                        "UPDATE sagas SET step_index=$1, context=$2, updated_at=now() WHERE id=$3",
                        step_index + 1, json.dumps(context), saga_id,
                    )
                except Exception as e:
                    log.error("step %s failed: %s — starting compensation", step["name"], e)
                    span.set_attribute("saga.failed", True)
                    await conn.execute(
                        "INSERT INTO saga_log (saga_id, step, action, result) VALUES ($1, $2, 'failed', $3)",
                        saga_id, step["name"], json.dumps({"error": str(e)}),
                    )
                    await conn.execute(
                        "UPDATE sagas SET status='COMPENSATING', updated_at=now() WHERE id=$1",
                        saga_id,
                    )

    row = await db_pool.fetchrow("SELECT status FROM sagas WHERE id=$1", saga_id)
    if row["status"] == "RUNNING":
        await advance(saga_id)
    elif row["status"] == "COMPENSATING":
        await compensate(saga_id)


async def compensate(saga_id: str):
    async with db_pool.acquire() as conn:
        row = await conn.fetchrow("SELECT step_index, context FROM sagas WHERE id=$1", saga_id)
        step_index = row["step_index"]
        context = json.loads(row["context"])

        for i in range(step_index - 1, -1, -1):
            step = STEPS[i]
            comp_name = step["compensate"]
            with tracer.start_as_current_span(f"saga.{comp_name}") as span:
                span.set_attribute("saga.id", saga_id)
                span.set_attribute("saga.compensate", comp_name)
                result = await execute_step(comp_name, context)
                await conn.execute(
                    "INSERT INTO saga_log (saga_id, step, action, result) VALUES ($1, $2, 'compensate', $3)",
                    saga_id, comp_name, json.dumps(result),
                )

        await conn.execute(
            "UPDATE sagas SET status='COMPENSATED', updated_at=now() WHERE id=$1",
            saga_id,
        )
    log.info("saga %s fully compensated", saga_id)


class SagaIn(BaseModel):
    order_id: str = Field(min_length=1)
    sku: str = Field(min_length=1)
    total: float = Field(gt=0)
    fail_shipping: bool = False


@asynccontextmanager
async def lifespan(app: FastAPI):
    global db_pool
    db_pool = await asyncpg.create_pool(DATABASE_URL, min_size=2, max_size=10)
    running = await db_pool.fetch("SELECT id FROM sagas WHERE status='RUNNING'")
    for row in running:
        log.info("resuming saga %s", row["id"])
        await advance(row["id"])
    log.info("saga-orchestrator started")
    yield
    await db_pool.close()


app = FastAPI(title="Saga Orchestrator", lifespan=lifespan)
FastAPIInstrumentor.instrument_app(app)


@app.get("/healthz")
async def healthz():
    return {"status": "ok"}


@app.post("/sagas", status_code=201)
async def create_saga(body: SagaIn):
    saga_id = str(uuid.uuid4())
    context = {"order_id": body.order_id, "sku": body.sku, "total": body.total}
    if body.fail_shipping:
        context["fail_shipping"] = True

    await db_pool.execute(
        "INSERT INTO sagas (id, status, step_index, context) VALUES ($1, 'RUNNING', 0, $2)",
        saga_id, json.dumps(context),
    )
    await advance(saga_id)

    row = await db_pool.fetchrow("SELECT id, status, step_index, context FROM sagas WHERE id=$1", saga_id)
    return {
        "id": row["id"],
        "status": row["status"],
        "step_index": row["step_index"],
        "context": json.loads(row["context"]),
    }


@app.get("/sagas/{saga_id}")
async def get_saga(saga_id: str):
    row = await db_pool.fetchrow("SELECT id, status, step_index, context FROM sagas WHERE id=$1", saga_id)
    if not row:
        return {"error": "not found"}, 404
    return {
        "id": row["id"],
        "status": row["status"],
        "step_index": row["step_index"],
        "context": json.loads(row["context"]),
    }


@app.get("/sagas/{saga_id}/log")
async def get_saga_log(saga_id: str):
    rows = await db_pool.fetch(
        "SELECT step, action, result, created_at FROM saga_log WHERE saga_id=$1 ORDER BY id",
        saga_id,
    )
    return [
        {"step": r["step"], "action": r["action"], "result": json.loads(r["result"]) if r["result"] else None}
        for r in rows
    ]

File: examples/17-sagas/cpp/saga-orchestrator/src/main.cpp
#include <drogon/drogon.h>
#include <libpq-fe.h>
#include <json/json.h>
#include <random>
#include <sstream>
#include <iomanip>
#include <vector>
#include <functional>
#include <thread>

using namespace drogon;

static std::string g_connStr;

static std::string generateId(const std::string& prefix) {
    static thread_local std::mt19937 rng(std::random_device{}());
    std::uniform_int_distribution<uint32_t> dist;
    std::ostringstream ss;
    ss << prefix << std::hex << std::setfill('0')
       << std::setw(8) << dist(rng);
    return ss.str();
}

static std::string generateUuid() {
    static thread_local std::mt19937 rng(std::random_device{}());
    std::uniform_int_distribution<uint32_t> dist;
    auto h = [&](int w) {
        std::ostringstream s;
        s << std::hex << std::setfill('0') << std::setw(w) << (dist(rng) >> (32 - w*4));
        return s.str();
    };
    return h(8) + "-" + h(4) + "-4" + h(3) + "-" + h(4) + "-" + h(8) + h(4);
}

struct PgConn {
    PGconn* conn = nullptr;
    PgConn() { conn = PQconnectdb(g_connStr.c_str()); }
    ~PgConn() { if (conn) PQfinish(conn); }
    bool ok() const { return conn && PQstatus(conn) == CONNECTION_OK; }
    PgConn(const PgConn&) = delete;
    PgConn& operator=(const PgConn&) = delete;
};

struct PgResult {
    PGresult* res = nullptr;
    PgResult(PGresult* r) : res(r) {}
    ~PgResult() { if (res) PQclear(res); }
    bool ok() const { return res && (PQresultStatus(res) == PGRES_TUPLES_OK || PQresultStatus(res) == PGRES_COMMAND_OK); }
    int rows() const { return res ? PQntuples(res) : 0; }
    std::string get(int r, int c) const { return PQgetvalue(res, r, c); }
    PgResult(const PgResult&) = delete;
    PgResult& operator=(const PgResult&) = delete;
};

static Json::Value parseJson(const std::string& s) {
    Json::Value root;
    Json::CharReaderBuilder builder;
    std::istringstream stream(s);
    Json::parseFromStream(builder, stream, &root, nullptr);
    return root;
}

static std::string toJsonString(const Json::Value& v) {
    Json::StreamWriterBuilder builder;
    builder["indentation"] = "";
    return Json::writeString(builder, v);
}

struct SagaStep {
    std::string name;
    std::string compensationName;
    std::function<Json::Value(const Json::Value&)> execute;
    std::function<Json::Value(const Json::Value&)> compensate;
};

static std::vector<SagaStep> g_steps = {
    {
        "charge_payment", "refund_payment",
        [](const Json::Value& ctx) {
            Json::Value r;
            r["payment_id"] = generateId("pay-");
            r["amount"] = ctx["total"].asDouble();
            return r;
        },
        [](const Json::Value& ctx) {
            Json::Value r;
            r["refunded"] = true;
            return r;
        }
    },
    {
        "reserve_stock", "release_stock",
        [](const Json::Value& ctx) {
            Json::Value r;
            r["reservation_id"] = generateId("rsv-");
            r["sku"] = ctx["sku"].asString();
            return r;
        },
        [](const Json::Value& ctx) {
            Json::Value r;
            r["released"] = true;
            return r;
        }
    },
    {
        "book_shipping", "cancel_shipping",
        [](const Json::Value& ctx) {
            if (ctx.isMember("fail_shipping") && ctx["fail_shipping"].asBool()) {
                throw std::runtime_error("shipping unavailable");
            }
            Json::Value r;
            r["shipment_id"] = generateId("shp-");
            return r;
        },
        [](const Json::Value& ctx) {
            Json::Value r;
            r["cancelled"] = true;
            return r;
        }
    }
};

static std::string escapeStr(const std::string& s) {
    std::string out;
    for (char c : s) {
        if (c == '\'') out += "''";
        else out += c;
    }
    return out;
}

static void logStep(PGconn* conn, const std::string& sagaId,
                    const std::string& step, const std::string& action,
                    const Json::Value& result) {
    std::string sql = "INSERT INTO saga_log (saga_id, step, action, result) VALUES ('"
        + escapeStr(sagaId) + "', '" + escapeStr(step) + "', '"
        + escapeStr(action) + "', '" + escapeStr(toJsonString(result)) + "')";
    PgResult r(PQexec(conn, sql.c_str()));
}

static void advance(const std::string& sagaId);
static void compensate(const std::string& sagaId);

static void advance(const std::string& sagaId) {
    PgConn db;
    if (!db.ok()) return;

    PQexec(db.conn, "BEGIN");

    std::string sql = "SELECT status, step_index, context FROM sagas WHERE id = '"
        + escapeStr(sagaId) + "' FOR UPDATE";
    PgResult row(PQexec(db.conn, sql.c_str()));
    if (!row.ok() || row.rows() == 0) { PQexec(db.conn, "ROLLBACK"); return; }

    std::string status = row.get(0, 0);
    int stepIdx = std::stoi(row.get(0, 1));
    Json::Value context = parseJson(row.get(0, 2));

    if (status != "RUNNING") { PQexec(db.conn, "ROLLBACK"); return; }

    int numSteps = static_cast<int>(g_steps.size());
    if (stepIdx >= numSteps) {
        std::string upd = "UPDATE sagas SET status = 'COMPLETED', updated_at = now() WHERE id = '"
            + escapeStr(sagaId) + "'";
        PQexec(db.conn, upd.c_str());
        PQexec(db.conn, "COMMIT");
        return;
    }

    auto& step = g_steps[stepIdx];
    try {
        Json::Value result = step.execute(context);
        context[step.name] = result;
        logStep(db.conn, sagaId, step.name, "execute", result);

        std::string upd = "UPDATE sagas SET step_index = " + std::to_string(stepIdx + 1)
            + ", context = '" + escapeStr(toJsonString(context))
            + "', updated_at = now() WHERE id = '" + escapeStr(sagaId) + "'";
        PQexec(db.conn, upd.c_str());
        PQexec(db.conn, "COMMIT");

        advance(sagaId);

    } catch (const std::exception& e) {
        Json::Value failResult;
        failResult["error"] = e.what();
        logStep(db.conn, sagaId, step.name, "failed", failResult);

        std::string upd = "UPDATE sagas SET status = 'COMPENSATING', updated_at = now() WHERE id = '"
            + escapeStr(sagaId) + "'";
        PQexec(db.conn, upd.c_str());
        PQexec(db.conn, "COMMIT");

        compensate(sagaId);
    }
}

static void compensate(const std::string& sagaId) {
    PgConn db;
    if (!db.ok()) return;

    PQexec(db.conn, "BEGIN");

    std::string sql = "SELECT status, step_index, context FROM sagas WHERE id = '"
        + escapeStr(sagaId) + "' FOR UPDATE";
    PgResult row(PQexec(db.conn, sql.c_str()));
    if (!row.ok() || row.rows() == 0) { PQexec(db.conn, "ROLLBACK"); return; }

    int stepIdx = std::stoi(row.get(0, 1));
    Json::Value context = parseJson(row.get(0, 2));

    // Compensate completed steps in reverse (step_index is next-to-run, so completed = 0..stepIdx-1)
    for (int i = stepIdx - 1; i >= 0; --i) {
        auto& step = g_steps[i];
        Json::Value result = step.compensate(context);
        logStep(db.conn, sagaId, step.compensationName, "compensate", result);
    }

    std::string upd = "UPDATE sagas SET status = 'COMPENSATED', updated_at = now() WHERE id = '"
        + escapeStr(sagaId) + "'";
    PQexec(db.conn, upd.c_str());
    PQexec(db.conn, "COMMIT");
}

static Json::Value getSaga(const std::string& sagaId) {
    PgConn db;
    if (!db.ok()) return Json::Value::null;

    std::string sql = "SELECT id, status, step_index, context FROM sagas WHERE id = '"
        + escapeStr(sagaId) + "'";
    PgResult row(PQexec(db.conn, sql.c_str()));
    if (!row.ok() || row.rows() == 0) return Json::Value::null;

    Json::Value saga;
    saga["id"] = row.get(0, 0);
    saga["status"] = row.get(0, 1);
    saga["step_index"] = std::stoi(row.get(0, 2));
    saga["context"] = parseJson(row.get(0, 3));
    return saga;
}

static Json::Value getSagaLog(const std::string& sagaId) {
    PgConn db;
    if (!db.ok()) return Json::Value(Json::arrayValue);

    std::string sql = "SELECT step, action, result FROM saga_log WHERE saga_id = '"
        + escapeStr(sagaId) + "' ORDER BY id";
    PgResult rows(PQexec(db.conn, sql.c_str()));
    if (!rows.ok()) return Json::Value(Json::arrayValue);

    Json::Value log(Json::arrayValue);
    for (int i = 0; i < rows.rows(); ++i) {
        Json::Value entry;
        entry["step"] = rows.get(i, 0);
        entry["action"] = rows.get(i, 1);
        entry["result"] = parseJson(rows.get(i, 2));
        log.append(entry);
    }
    return log;
}

static void resumeRunningSagas() {
    PgConn db;
    if (!db.ok()) {
        LOG_ERROR << "Cannot connect to DB for saga resume";
        return;
    }

    PgResult rows(PQexec(db.conn, "SELECT id FROM sagas WHERE status = 'RUNNING'"));
    if (!rows.ok()) return;

    for (int i = 0; i < rows.rows(); ++i) {
        std::string id = rows.get(i, 0);
        LOG_INFO << "Resuming saga: " << id;
        advance(id);
    }
}

int main() {
    const char* dbUrl = std::getenv("DATABASE_URL");
    g_connStr = dbUrl ? dbUrl : "postgres://appuser:apppass@localhost:5432/appdb";

    app().setLogLevel(trantor::Logger::kInfo);
    app().setUploadPath("/tmp");

    app().registerHandler("/healthz",
        [](const HttpRequestPtr&, std::function<void(const HttpResponsePtr&)>&& callback) {
            Json::Value resp;
            resp["status"] = "ok";
            callback(HttpResponse::newHttpJsonResponse(resp));
        },
        {Get});

    app().registerHandler("/sagas",
        [](const HttpRequestPtr& req, std::function<void(const HttpResponsePtr&)>&& callback) {
            auto body = req->getJsonObject();
            if (!body) {
                auto resp = HttpResponse::newHttpJsonResponse(Json::Value("invalid json"));
                resp->setStatusCode(k400BadRequest);
                callback(resp);
                return;
            }

            std::string sagaId = generateUuid();
            Json::Value context = *body;

            PgConn db;
            if (!db.ok()) {
                auto resp = HttpResponse::newHttpJsonResponse(Json::Value("db error"));
                resp->setStatusCode(k500InternalServerError);
                callback(resp);
                return;
            }

            std::string sql = "INSERT INTO sagas (id, status, step_index, context) VALUES ('"
                + escapeStr(sagaId) + "', 'RUNNING', 0, '" + escapeStr(toJsonString(context)) + "')";
            PgResult r(PQexec(db.conn, sql.c_str()));

            advance(sagaId);

            Json::Value saga = getSaga(sagaId);
            auto resp = HttpResponse::newHttpJsonResponse(saga);
            resp->setStatusCode(k201Created);
            callback(resp);
        },
        {Post});

    app().registerHandler("/sagas/{saga_id}",
        [](const HttpRequestPtr& req, std::function<void(const HttpResponsePtr&)>&& callback,
           const std::string& sagaId) {
            Json::Value saga = getSaga(sagaId);
            if (saga.isNull()) {
                auto resp = HttpResponse::newHttpJsonResponse(Json::Value("not found"));
                resp->setStatusCode(k404NotFound);
                callback(resp);
                return;
            }
            callback(HttpResponse::newHttpJsonResponse(saga));
        },
        {Get});

    app().registerHandler("/sagas/{saga_id}/log",
        [](const HttpRequestPtr& req, std::function<void(const HttpResponsePtr&)>&& callback,
           const std::string& sagaId) {
            callback(HttpResponse::newHttpJsonResponse(getSagaLog(sagaId)));
        },
        {Get});

    // Resume any incomplete sagas after startup
    app().getLoop()->runAfter(2.0, []() {
        std::thread([]() { resumeRunningSagas(); }).detach();
    });

    LOG_INFO << "Saga orchestrator starting on 0.0.0.0:8080";
    app().addListener("0.0.0.0", 8080).run();
}
